File: proj.android/src/cn/robust/mastermind/AccessibilityHelper.java
package cn.robust.mastermind;

import java.lang.ref.WeakReference;

import android.annotation.TargetApi;
import android.graphics.Rect;
import android.os.Build;
import android.support.v4.view.ViewCompat;
import android.util.Log;
import android.view.accessibility.AccessibilityEvent;

public class AccessibilityHelper {
	
	private static int sGameWith = 480;
	private static int sGameHeight = 720;
	
	private static int sWidth;
	private static int sHeight;
	
	private static BaseSceneHelper mMenuRef;
	private static BaseSceneHelper mPlayRef;
	private static BaseSceneHelper mOverRef;
	private static BaseSceneHelper mHelpRef;
	private static BaseSceneHelper mCurRef;
	
	private static WeakReference<AccessibilityGameView> mGameViewRef;
	
	public static void setGameView(AccessibilityGameView view){
		mGameViewRef = new WeakReference<AccessibilityGameView>(view);
		
		mMenuRef = new BaseSceneHelper(view);
		mPlayRef = new BaseSceneHelper(view);
		mOverRef = new BaseSceneHelper(view);
		mHelpRef = new BaseSceneHelper(view);
	}
	
	public static void setScreen(int w, int h){
		sWidth = w;
		sHeight = h;
		
		Log.e("hjy", "setScreen w:" + w + " h:" + h);
	}
	
	public static int getWidth(){
		return sWidth;
	}
	
	public static int getHeight(){
		return sHeight;
	}
	
	private static int getScreenX(int x){
		int ret = x * sWidth / sGameWith;
		return ret;
	}
	
	private static int getScreenY(int y){
		// 针对不同的拉伸方式要有不同的转换，这里是kResolutionShowAll
		int realHeight = sGameHeight * sWidth / sGameWith;
		int blackHeight = (sHeight - realHeight) / 2;
		int y1 = sGameHeight - y;
		int ret = y1 * realHeight / sGameHeight + blackHeight;
		return ret;
	}
	
	public static Rect getScreenRect(Rect r){
		int left = getScreenX(r.left);
		int right = getScreenX(r.right);
		int top = getScreenY(r.top);
		int bottom = getScreenY(r.bottom);
		
		return new Rect(left, top, right, bottom);
	}
	
	public static void onMenuSceneLoad(int scene){
		Log.e("hjy", "onMenuSceneLoad " + scene);
		
		if(mGameViewRef.get() != null && mMenuRef != null && 
				mPlayRef != null && mOverRef != null){
			switch (scene) {
			case 0:
				ViewCompat.setAccessibilityDelegate(mGameViewRef.get(), mMenuRef);
				handleNewScene(mMenuRef);
				break;
			case 1:
				ViewCompat.setAccessibilityDelegate(mGameViewRef.get(), mPlayRef);
				handleNewScene(mPlayRef);
				break;
			case 2:
				ViewCompat.setAccessibilityDelegate(mGameViewRef.get(), mOverRef);
				handleNewScene(mOverRef);
				break;
			case 3:
				ViewCompat.setAccessibilityDelegate(mGameViewRef.get(), mHelpRef);
				handleNewScene(mHelpRef);
			}
		}
	}
	
	private static void handleNewScene(BaseSceneHelper newScene){
		if(mCurRef != null ){
			mCurRef.destroyScene();
		}
		mCurRef = newScene;
		mGameViewRef.get().setCurSceneHelper(newScene);
		
		if(android.os.Build.VERSION.SDK_INT >= 16){
			mGameViewRef.get().postDelayed(new Runnable() {
				
				@TargetApi(Build.VERSION_CODES.JELLY_BEAN)
				@Override
				public void run() {
					mGameViewRef.get().sendAccessibilityEvent(AccessibilityEvent.TYPE_VIEW_ACCESSIBILITY_FOCUS_CLEARED);
				}
			}, 100);
		}
	}
	
	public static void addMenuSceneRect(int i, String d, int l, int r, int t, int b){
		int sl = getScreenX(l);
		int sr = getScreenX(r);
		int st = getScreenY(b);
		int sb = getScreenY(t);
		AccessibilityItem item = new AccessibilityItem(i, d, sl, sr, st, sb);
		Log.e("hjy", "addMenuSceneRect " + item.toString());
		if(mMenuRef != null){
			mMenuRef.addAccessibilityItem(item);
		}
	}
	
	public static void addPlaySceneRect(int i, String d, int l, int r, int t, int b){
		Log.e("hjy", "addPlaySceneRect rect:" + l + "," + r + "," + t + "," + b);
		int sl = getScreenX(l);
		int sr = getScreenX(r);
		int st = getScreenY(b);
		int sb = getScreenY(t);
		AccessibilityItem item = new AccessibilityItem(i, d, sl, sr, st, sb);
		//Log.e("hjy", "addPlaySceneRect " + item.toString());
		if(mPlayRef != null){
			mPlayRef.addAccessibilityItem(item);
		}
	}
	
	public static void addSceneRect(int i, String d, int l, int r, int t, int b){
		int sl = getScreenX(l);
		int sr = getScreenX(r);
		int st = getScreenY(b);
		int sb = getScreenY(t);
		AccessibilityItem item = new AccessibilityItem(i, d, sl, sr, st, sb);
		Log.e("hjy", "addSceneRect " + item.toString());
		if(mCurRef != null){
			mCurRef.addAccessibilityItem(item);
		}
	}
	
	@TargetApi(Build.VERSION_CODES.JELLY_BEAN) 
	public static void annouceStepResult(String result){
		if(mGameViewRef.get() != null){
			if(android.os.Build.VERSION.SDK_INT >= 16){
			    AccessibilityEvent event = AccessibilityEvent.obtain(AccessibilityEvent.TYPE_ANNOUNCEMENT);
			    event.setPackageName(mGameViewRef.get().getContext().getPackageName());
			    event.setClassName(mGameViewRef.get().getClass().getName());
			    event.setSource(mGameViewRef.get());
			    event.getText().add(result);
			    mGameViewRef.get().getParent().requestSendAccessibilityEvent(mGameViewRef.get(), event);
			}
		}
	}
	
	public static void updateAccessibilityItem(int i, String desc){
		if(mCurRef != null){
			mCurRef.updateAccessibilityItem(i, desc);
	}
	}
}

File: proj.ios/AccessibilityImpl/AccessibilityHelper.m
//
//  AccessibilityHelper.m
//  Mole Guess
//
//  Created by Freeman on 2016/10/22.
//
//

#import "AccessibilityHelper.h"

@implementation AccessibilityHelper

static AccessibilityHelper * instance = nil;

+(AccessibilityHelper*) getInstance{
    @synchronized(self) {
        if (instance == nil) {
            instance = [[self alloc] init];
            [instance onInited];
        }
    }
    return instance;
}

-(void) onInited{
    CGRect rect = [[UIScreen mainScreen] bounds];
    sWidth = rect.size.width;
    sHeight = rect.size.height;
    
    mMenuRef = [[BaseSceneHelper alloc] init];
    mPlayRef = [[BaseSceneHelper alloc] init];
    mOverRef = [[BaseSceneHelper alloc] init];
    mHelpRef = [[BaseSceneHelper alloc] init];
}

-(void) onMenuSceneLoad:(int)scene{
    switch (scene) {
        case 0:
            [self handleNewScene:mMenuRef];
            break;
        case 1:
            [self handleNewScene:mPlayRef];
            break;
        case 2:
            [self handleNewScene:mOverRef];
            break;
        case 3:
            [self handleNewScene:mHelpRef];
            break;
        default:
            NSLog(@"onMenuSceneLoad %d error", scene);
            break;
    }

}

-(void) handleNewScene:(BaseSceneHelper *)newScene{
    if(mCurRef != NULL){
        [mCurRef destroyScene];
    }
    if(mGameView){
        [mGameView onSceneDestroy];
    }
    
    mCurRef = newScene;
    
    if(mGameView){
        [mGameView onSceneChange:newScene];
    }
}

-(void) addMenuSceneRect:(NSString *)d left:(int)l right:(int)r top:(int)t bottom:(int)b{
    int sl = [self getScreenX:l];
    int sr = [self getScreenX:r];
    int st = [self getScreenY:b];
    int sb = [self getScreenY:t];
    AccessibilityItem * item = [[AccessibilityItem alloc] init];
    CGRect rect = [self getScreenRect:sl right:sr top:st bottom:sb];
    [item initWithDesc:d withRect:rect];
    if(mMenuRef != NULL){
        [mMenuRef addAccessibilityItem:item];
    }
}

-(void) addPlaySceneRect:(NSString *)d left:(int)l right:(int)r top:(int)t bottom:(int)b{
    int sl = [self getScreenX:l];
    int sr = [self getScreenX:r];
    int st = [self getScreenY:b];
    int sb = [self getScreenY:t];
    AccessibilityItem * item = [[AccessibilityItem alloc] init];
    CGRect rect = [self getScreenRect:sl right:sr top:st bottom:sb];
    [item initWithDesc:d withRect:rect];
    if(mPlayRef != NULL){
        [mPlayRef addAccessibilityItem:item];
    }
}

-(void) addSceneRect:(NSString *)d left:(int)l right:(int)r top:(int)t bottom:(int)b{
    int sl = [self getScreenX:l];
    int sr = [self getScreenX:r];
    int st = [self getScreenY:b];
    int sb = [self getScreenY:t];
    AccessibilityItem * item = [[AccessibilityItem alloc] init];
    CGRect rect = [self getScreenRect:sl right:sr top:st bottom:sb];
    [item initWithDesc:d withRect:rect];
    if(mCurRef != NULL){
        [mCurRef addAccessibilityItem:item];
    }
}

-(void) annouceStepResult:(NSString *)d{
    [self performSelector:@selector(annouceDelayImpl:) withObject:[NSString stringWithString:d] afterDelay:1.0f];
}

-(void) annouceDelayImpl:(NSString *)d{
    // 主动发声
    UIAccessibilityPostNotification(UIAccessibilityAnnouncementNotification, d);
}

-(void) setGameView:(AccessibilityView *)view{
    mGameView = view;
}

-(int) getScreenY:(int)y{
    int rH = GAME_HEIGHT * sWidth / GAME_WIDTH;
    int bH = (sHeight - rH) / 2;
    int y1 = GAME_HEIGHT - y;
    return (y1 * rH / GAME_HEIGHT + bH);
}

-(int) getScreenX:(int)x{
    return x * sWidth / GAME_WIDTH;
}

-(CGRect) getScreenRect:(int)l right:(int)r top:(int)t bottom:(int)b{
   // NSLog(@"%d,%d,%d,%d",l, r, t, b);
    return CGRectMake(l, t, r - l, b - t);
}

@end
